core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java

```diff
@@ -177,192 +166,313 @@ public DefaultRequestLog(RequestContext ctx, ContentPreviewerFactory requestCont
     }
 
-    private static int getterFlags(Iterable<RequestLogAvailability> availabilities) {
-        int flags = 0;
-        for (RequestLogAvailability a : availabilities) {
-            flags |= a.getterFlags();
+    // Methods required for adding children.
+
+    @Override
+    public void addChild(RequestLogAccess child) {
+        checkState(!hasLastChild, "last child is already added");
+        requireNonNull(child, "child");
+        if (children == null) {
+            // first child's all request-side logging events are propagated immediately to the parent
+            children = new ArrayList<>();
+            propagateRequestSideLog(child);
         }
-        return flags;
+        children.add(child);
+    }
+
+    private void propagateRequestSideLog(RequestLogAccess child) {
+        child.partialFuture(RequestLogProperty.SESSION, RequestLogProperty.REQUEST_START_TIME)
+             .thenAccept(log -> {
+                 final ClientConnectionTimings timings = ClientConnectionTimings.get(log);
+                 if (timings != null) {
+                     timings.setTo(this);
+                 }
+                 startRequest0(log.channel(), log.sessionProtocol(), null,
+                               log.requestStartTimeNanos(), log.requestStartTimeMicros(), true);
+             });
+        child.partialFuture(RequestLogProperty.SCHEME)
+             .thenAccept(log -> serializationFormat(log.scheme().serializationFormat()));
+        child.partialFuture(RequestLogProperty.REQUEST_FIRST_BYTES_TRANSFERRED_TIME)
+             .thenAccept(log -> {
+                 final Long timeNanos = log.requestFirstBytesTransferredTimeNanos();
+                 if (timeNanos != null) {
+                     requestFirstBytesTransferred(timeNanos);
+                 }
+             });
+        child.partialFuture(RequestLogProperty.REQUEST_HEADERS)
+             .thenAccept(log -> requestHeaders(log.requestHeaders()));
+        child.partialFuture(RequestLogProperty.REQUEST_CONTENT)
+             .thenAccept(log -> requestContent(log.requestContent(), log.rawRequestContent()));
+        child.requestCompleteFuture().thenAccept(log -> {
+            requestLength(log.requestLength());
+            requestContentPreview(log.requestContentPreview());
+            requestTrailers(log.requestTrailers());
+            endRequest0(log.requestCause(), log.requestEndTimeNanos());
+        });
     }
 
     @Override
-    public RequestContext context() {
-        return ctx;
+    public void endResponseWithLastChild() {
+        checkState(!hasLastChild, "last child is already added");
+        checkState(children != null && !children.isEmpty(), "at least one child should be already added");
+        hasLastChild = true;
+        final RequestLogAccess lastChild = children.get(children.size() - 1);
+        propagateResponseSideLog(lastChild.partial());
+    }
+
+    private void propagateResponseSideLog(RequestLog lastChild) {
```

Comment:
By the way when I was reading this code I was confused by response checking availability but not request. I'm assuming the request will never be available, but adding a comment why would help

Reply:
Added some comments to both `propagate{Request,Response}SideLog()`